ouroboros-consensus/docs/FastLaneSketch.md

```diff
@@ -0,0 +1,152 @@
+* relevant background thread https://github.com/input-output-hk/ouroboros-network/blob/b0d95f33314e0562b6eea55455f1847efb85f8a4/ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/Background.hs#L523-L536
+* chain selection logic https://github.com/input-output-hk/ouroboros-network/blob/b0d95f33314e0562b6eea55455f1847efb85f8a4/ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs
+
+# Initial implementation plan
+
+I think I see a lightweight means of extending the above system to "rush"
+through the latest block and to do so only when it would matter.
+
+* Extend the ChainDB API with a part that notifies subscribed threads whenever
+  a new most-desirable-block-ever-seen is added.
+* Define a new mini protocol named FastLane as a pared down ChainSync.
+    * It has two messages: `MsgRequestNextFast`, `MsgFast`
+    * The intended semantics is that the server sends `MsgFast hdr` to the
+      client as soon as the server downloads a new most-desirable block.
+    * Notably, the server sends `MsgFast` _before_ attempting to validate the
+      new block.
+* Expand the input to the block fetch decision logic so that it includes the
+  latest header received via FastLane.
+* Crucially, peers are to be forgiven for sending us an invalid block this way.
+
+That's the big picture. Some more details follow.
+
+* Unlike ChainSync, The FastLane client ignores the header if it doesn't
+  directly extend the node's selected chain.
+    * I'm tempted to make it more liberal, but that would involve extra
+      complexity, along the line of checking whether all the blocks between the
+      new header and its intersection with our selection are already in the
+      VolDB.
+    * Notably, if pipelining has been working, then the simpler check should
+      suffice.
+* Like ChainSync, if the header is valid, then the FastLane client makes its
+  available to the BlockFetch decision logic.
+    * The FastLane client disconnects from the peer if the header extends the
+      node's selection chain but is invalid.
+* The FastLane server does not reply to `MsgRequestNextFast` until it has a
+  _new_ most-desirable-block-ever-seen.
+    * The `MsgFast` timeout is comparable to the ChainSync timeout after
+      receiving `MsgAwaitReply`.
+    * Every `MsgFast` should be strictly preferable to the previous `MsgFast`
+      that peer sent.
+* Like ChainSync, FastLane is pull-based, so we can't get overwhelmed.
+* Like ChainSync, FastLane is also pipelined, so that we can hide some latency.
+* If a peer offered a (valid) header for an invalid block via ChainSync we
+  disconnect. If the peer only offered instead that header via FastLane, then
+  we don't disconnect.
+
+# Doubts
+
+* Should FastLane just send the block? Is the extra RTT worthwhile? If it just
```

Comment:
I don't think so. I think we should mirror the existing primitive which is predicated on headers.